[Doc] Update tracing quickstart to include Typescript SDK (#16630)
mlflow/mlflow: docs/docs/genai/tracing/quickstart/python-openai.mdx

> import Tabs from '@theme/Tabs';

Author: Interesting. Directing `tracing` to this page is expected, but the side bar shouldn't open both. Lemme check if there is any option in docusaurus to prevent this.

> import Tabs from '@theme/Tabs';

Author: Hmm the redirect hack doesn't work :(

> The redirect plugin is not supposed to override existing files. I'll remove this link temporary and update getting-started page in a follow-up PR.

# Tracing Quickstart (Python)

:::info

The TypeScript/JavaScript quickstart is available <ins>[here](/genai/tracing/quickstart/typescript-openai)</ins>.

:::

This quickstart guide will walk you through setting up a simple GenAI application with MLflow Tracing. In less than 10 minutes, you'll enable tracing, run a basic application, and explore the generated traces in the MLflow UI.

## Prerequisites

Install the required packages by running the following command:

```bash
pip install --upgrade mlflow openai>=1.0.0
```

:::info

The code example in this guide uses OpenAI SDK, however, the contents of this guide are applicable to any other LLM providers, such as Anthropic, Google, Bedrock, and more.

:::

## Step 1: Set up your environment

### Connect to MLflow

MLflow logs Traces in a tracking server. Connect your local environment to the tracking server by one of the following methods.

<TabsWrapper>
<Tabs>
<TabItem value="local" label="Local MLflow" default>

For the fastest setup, you can run MLflow locally:

```bash
# Start MLflow tracking server locally
mlflow ui

# This will start the server at http://127.0.0.1:5000
```

</TabItem>

> <TabItem value="remote" label="Remove MLflow Server">

Reviewer: ```suggestion
<TabItem value="remote" label="Remote MLflow Server">
```

```python
import os

# Use different env variable when using a different LLM provider
os.environ["OPENAI_API_KEY"] = "your-api-key-here"  # Replace with your actual API key
```

## Step 2: Trace a single LLM call

Let's start with a simple example of tracing a single LLM call. We first enable automatic tracing for OpenAI API calls by calling `mlflow.openai.autolog()`. After that, every calls to OpenAI API will generate as a trace span, capturing the input, output, latency, token counts, and other metadata.

```python
import mlflow
from openai import OpenAI

# Enable automatic tracing for all OpenAI API calls
mlflow.openai.autolog()

# Instantiate the OpenAI client
client = OpenAI()

# Invoke chat completion API
response = client.chat.completions.create(
    model="o4-mini",
    messages=[
        {"role": "system", "content": "You are a helpful weather assistant."},
        {"role": "user", "content": "What's the weather like in Seattle?"},
    ],
)
```

After running the code above, go to the MLflow UI and select "Traces" tab. It should show the newly created trace.

<ImageBox src="/images/llms/tracing/quickstart/single-openai-trace-list.png" alt="Single Trace" />

> The table view shows primary metadata of the trace, such as the trace ID, execution duration, token count, source system, status. You can add/remove columns by selecting the columns in the drop down. By clicking on the request row (the linked request text), you can view the detailed spans in the trace.

Reviewer 2: ```suggestion
The table view shows the primary metadata of the trace, such as the trace ID, execution duration, token count, source system and status. You can add or remove displayed columns by selecting the columns in the drop down. By clicking on the request row (the linked request text), you can view the detailed spans in the trace.
```

<ImageBox src="/images/llms/tracing/quickstart/single-openai-trace-detail.png" alt="Single Trace Detail" width="80%" />

> The "Chat" view in the above screenshot shows the full chat messages exchanged between the user and the model. By clicking other table like "Inputs / Outputs", "Attributes", you can see different aspects of the trace, such as the raw input payload, token usage breakdown, and more.

Reviewer 2: ```suggestion
The "Chat" view in the above screenshot shows the full chat messages exchanged between the user and the model. By clicking other tables such as "Inputs / Outputs" or "Attributes", you can see different aspects of the trace, including the raw input payload, token usage breakdown, and more.
```

## Step 3: Trace a tool calling agent

> Next, let's add a bit more complexity to the application. To get the real-time weather information, we will use an external weather API as a tool. The application will include a tool calling flow, not only a simple LLM call. To instrument that custom Python flow, we will use the <APILink fn="mlflow.trace">`@mlflow.trace`</APILink> decorator. The powerful decorator generates a span for the function call and automatically capture the function inputs, outputs, latency, and exceptions.

Reviewer 2: ```suggestion
Next, let's add a bit more complexity to the application. To get the real-time weather information, we will use an external weather API as a tool. The application will include a tool calling flow, not only a simple LLM call. To instrument that custom Python flow, we will use the <APILink fn="mlflow.trace">`@mlflow.trace`</APILink> decorator. Use of this simple decorator generates a span for the function call and automatically captures the function inputs, outputs, latency, and exceptions.
```

```python
tools = [
    {
        "type": "function",
        "function": {
            "name": "get_weather",
            "description": "Get current temperature for provided coordinates in celsius.",
            "parameters": {
                "type": "object",
                "properties": {
                    "latitude": {"type": "number"},
                    "longitude": {"type": "number"},
                },
                "required": ["latitude", "longitude"],
                "additionalProperties": False,
            },
            "strict": True,
        },
    }
]
```

Lastly, define a simple flow that first ask LLM to get instruction for calling the tool, invoke the tool function, and return the result to the LLM.

```python
import json

# Define a simple tool calling agent
@mlflow.trace(span_type=SpanType.AGENT)
def run_tool_agent(question: str):
    messages = [{"role": "user", "content": question}]

    # Invoke the model with the given question and available tools
    response = client.chat.completions.create(
        model="o4-mini",
        messages=messages,
        tools=tools,
    )
    ai_msg = response.choices[0].message
    messages.append(ai_msg)

>     # If the model request tool call(s), invoke the function with the specified arguments

Reviewer 2: ```suggestion
    # If the model requests tool call(s), invoke the function with the specified arguments
```
